htlcswitch/link_test.go

```diff
@@ -536,6 +400,102 @@ func testChannelLinkMultiHopPayment(t *testing.T,
 	}
 }
 
+// TestChannelLinkCancelFullCommitment tests the ability for links to cancel
+// forwarded HTLCs once all of their commitment slots are full.
+func TestChannelLinkCancelFullCommitment(t *testing.T) {
+	t.Parallel()
+
+	channels, cleanUp, _, err := createClusterChannels(
+		btcutil.SatoshiPerBitcoin*3,
+		btcutil.SatoshiPerBitcoin*5)
+	if err != nil {
+		t.Fatalf("unable to create channel: %v", err)
+	}
+	defer cleanUp()
+
+	n := newTwoHopNetwork(
+		t, channels.aliceToBob, channels.bobToAlice, testStartingHeight,
+	)
+	if err := n.start(); err != nil {
+		t.Fatal(err)
+	}
+	defer n.stop()
+
+	// Fill up the commitment from Alice's side with 20 sat payments.
+	count := (input.MaxHTLCNumber / 2)
+	amt := lnwire.NewMSatFromSatoshis(20000)
+
+	htlcAmt, totalTimelock, hopsForwards := generateHops(amt,
+		testStartingHeight, n.bobChannelLink)
+
+	firstHop := n.aliceChannelLink.ShortChanID()
+
+	// Create channels to buffer the preimage and error channels used in
+	// making the preliminary payments.
+	preimages := make([]lntypes.Preimage, count)
+	aliceErrChan := make(chan chan error, count)
+
+	var wg sync.WaitGroup
+	for i := 0; i < count; i++ {
+		preimages[i] = lntypes.Preimage{byte(i >> 8), byte(i)}
+
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+
+			errChan := n.makeHoldPayment(
+				n.aliceServer, n.bobServer, firstHop,
+				hopsForwards, amt, htlcAmt, totalTimelock,
+				preimages[i],
+			)
+			aliceErrChan <- errChan
+		}(i)
+	}
+
+	// Wait for Alice to finish filling her commitment.
+	wg.Wait()
+	close(aliceErrChan)
+
+	// Now make an additional payment from Alice to Bob, this should be
+	// canceled because the commitment in this direction is full.
+	err = <-makePayment(
+		n.aliceServer, n.bobServer, firstHop, hopsForwards, amt,
+		htlcAmt, totalTimelock,
+	).err
+	if err == nil {
+		t.Fatalf("overflow payment should have failed")
+	}
+	lerr, ok := err.(*LinkError)
+	if !ok {
+		t.Fatalf("expected LinkError, got: %T", err)
+	}
+
+	msg := lerr.WireMessage()
+	if _, ok := msg.(*lnwire.FailTemporaryChannelFailure); !ok {
+		t.Fatalf("expected TemporaryChannelFailure, got: %T", msg)
+	}
+
+	// Now, settle all htlcs held by bob and clear the commitment of htlcs.
+	for _, preimage := range preimages {
+		preimage := preimage
+
+		err = wait.NoError(func() error {
```

Comment:
Document why `wait.NoError`? Is it because the htlcs may not have arrived yet? Subscribing to invoice events would have been cleaner than polling perhaps.

Reply:
done